### PR TITLE
[MLIR] Fix incorrect memref::DimOp canonicalization, add tensor::DimOp canonicalization

### PR DESCRIPTION
**Pull request by sahas3**
The current canonicalization of `memref.dim` operating on the result of `memref.reshape` into `memref.load` is incorrect as it doesn't check whether the `index` operand of `memref.dim` dominates the source `memref.reshape` op. It always introduces `memref.load` right after `memref.reshape` to ensure the `memref` is not mutated before the `memref.load` call. As a result, the following error is observed:

```
$> mlir-opt --canonicalize input.mlir

func.func @reshape_dim(%arg0: memref<*xf32>, %arg1: memref<?xindex>, %arg2: index) -> index {
    %c4 = arith.constant 4 : index
    %reshape = memref.reshape %arg0(%arg1) : (memref<*xf32>, memref<?xindex>) -> memref<*xf32>
    %0 = arith.muli %arg2, %c4 : index
    %dim = memref.dim %reshape, %0 : memref<*xf32>
    return %dim : index
  }
```

results in:

```
dominator.mlir:22:12: error: operand #1 does not dominate this use
    %dim = memref.dim %reshape, %0 : memref<*xf32>
           ^
dominator.mlir:22:12: note: see current operation: %1 = "memref.load"(%arg1, %2) <{nontemporal = false}> : (memref<?xindex>, index) -> index
dominator.mlir:21:10: note: operand defined here (op in the same block)
    %0 = arith.muli %arg2, %c4 : index
```

Properly fixing this issue requires a dominator analysis which is expensive to run within a canonicalization pattern. So, this patch moves the canonicalization pattern to `tensor.dim`. Since tensors are immutable we don't need to worry about where to introduce the `tensor.extract` call after canonicalization.

### CODE REVIEW

**Comment by joker-eph** [comment]
LGTM overall, but should we also have the negative tests you have on the memref::DimOp?

**Comment by MaheshRavishankar** [comment]
I am not sure that removing the `memref.dim` canonicalization here is the right answer. The canonicalization itself seems fine apart from it inserting operation after the op.

**Review by MaheshRavishankar — DISMISSED** [review]
I don't think the memrefs canonicalization needs to be removed

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp [hunk 1]
```diff
@@ -1069,39 +1069,6 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
...
-
-    if (!reshape)
-      return failure();
-
-    // Place the load directly after the reshape to ensure that the shape memref
-    // was not mutated.
-    rewriter.setInsertionPointAfter(reshape);
-    Location loc = dim.getLoc();
-    Value load =
-        rewriter.create<LoadOp>(loc, reshape.getShape(), dim.getIndex());
```
Comment (matthias-springer) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1090]: To avoid the dominance error: Instead of full `DominanceInfo` you could do a cheaper check here that:
* `dim.getIndex()` is defined in a parent block of `reshape`.
* Or: `dim.getIndex()` is defined in the same block as `reshape` but before `reshape`.

The annoying part is that `dim.getIndex()` could be a block argument, so there's one more special case. I've have into use cases where I needed this kind of check (defined in parent op or in the same block before a given op) many times, so we could add a such a helper function to `Dominance.h`.

Comment (joker-eph) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1090]: > Or: dim.getIndex() is defined in the same block as reshape but before reshape.

If you do that, then you can just move the insertion point to the later of the index or the reshape I think.

Comment (sahas3) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1090]: >     * `dim.getIndex()` is defined in a parent block of `reshape`.
This will require recursing through all parent blocks of `reshape` -- is that cheap enough to be part of canonicalization? Should this be it's own pass itself instead of a canonicalization pattern?

Comment (matthias-springer) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1090]: IMO it's cheap enough. You can do something along the lines of `dim.getIndex().getParentRegion()->isProperAncestor(reshape->getParentRegion())`.

Comment (sahas3) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1090]: Oh great, that does what I was thinking. Still learning about the codebase. Thanks!

However, with the `tensor.dim` canonicalization, is there a requirement for canonicalizing `memref.dim` for the same pattern too?

Comment (matthias-springer) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1090]: `tensor.dim` and `memref.dim` used to be the same operation and we split it into two separate ops at some point. One the one hand, op canonicalizations should be defined in the dialect of the op (i.e., splitting it up would make sense). On the other hand, in this case it means duplicating some code. Personally, I'd rather have two patterns, one in the tensor dialect and one in the memref dialect, so that the two dialects are decoupled as much as possible.

File: mlir/lib/Dialect/Linalg/Transforms/Loops.cpp [hunk 2]
```diff
@@ -317,7 +317,6 @@ static void lowerLinalgToLoopsImpl(Operation *enclosingOp) {
   MLIRContext *context = enclosingOp->getContext();
   RewritePatternSet patterns(context);
   patterns.add<LinalgRewritePattern<LoopType>>(context);
-  memref::DimOp::getCanonicalizationPatterns(patterns, context);
```
Comment (sahas3) [inline mlir/lib/Dialect/Linalg/Transforms/Loops.cpp:320]: Revert this change.

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp [hunk 3]
```diff
@@ -1069,6 +1069,52 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
...
+      if (auto *definingOp = dim.getIndex().getDefiningOp()) {
+        if (reshape->isBeforeInBlock(definingOp))
+          return rewriter.notifyMatchFailure(
+              dim,
+              "dim.getIndex is not defined before reshape in the same block.");
+      } // else dim.getIndex is a block argument to reshape->getBlock
+    } else if (!dim.getIndex().getParentRegion()->isProperAncestor(
+                   reshape->getParentRegion()))
+      return rewriter.notifyMatchFailure(
+          dim, "dim.getIndex does not dominate reshape.");
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1096]: You need braces here on the else because braces exists on the if

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp [hunk 4]
```diff
@@ -1069,6 +1069,52 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
...
+
+  LogicalResult matchAndRewrite(DimOp dim,
+                                PatternRewriter &rewriter) const override {
+    auto reshape = dim.getSource().getDefiningOp<ReshapeOp>();
+
+    if (!reshape)
+      return rewriter.notifyMatchFailure(
+          dim, "Dim op is not defined by a reshape op.");
+
+    if (dim.getIndex().getParentBlock() == reshape->getBlock()) {
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1086]: This likely deserve some comments, maybe with pseudo code, about what are we achieving here.

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp [hunk 5]
```diff
@@ -1069,6 +1069,52 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
...
+      if (auto *definingOp = dim.getIndex().getDefiningOp()) {
+        if (reshape->isBeforeInBlock(definingOp))
+          return rewriter.notifyMatchFailure(
+              dim,
+              "dim.getIndex is not defined before reshape in the same block.");
+      } // else dim.getIndex is a block argument to reshape->getBlock
+    } else if (!dim.getIndex().getParentRegion()->isProperAncestor(
```
Comment (joker-eph) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1093]: A shortcut condition before this one can be `reshape->getParentBlock() == dim->getParentBlock()`

**Review by joker-eph — COMMENTED** [review]
The tensor dimOp canonicalization seemed valuable, are you bringing this another PR?

**Comment by sahas3** [comment]
> The tensor dimOp canonicalization seemed valuable, are you bringing this another PR?

`tensor.dim` canonicalization is still part of this PR: https://github.com/llvm/llvm-project/pull/84225/files/e50232affc5a111db92b63df30031031b6ce7bc4#diff-025c342b2a76f2edaae5557ff3e81b0b9ef9f47107359e78d85654581318d154 

Are you suggesting it should be it's own PR?

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp (hunk 4 above)
Comment (sahas3) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1086]: I was hoping that the messages in the `notifyMatchFailure` will inform the intention of the two conditions. I've added some more comments to make the intention clearer.

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp (hunk 5 above)
Comment (sahas3) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1093]: I am not following how this check will help. Can you elaborate?

**Comment by joker-eph** [comment]
> > The tensor dimOp canonicalization seemed valuable, are you bringing this another PR?
> 
> @joker-eph `tensor.dim` canonicalization is still part of this PR

Nevermind: GitHub was showing me the diff since the last review (or just for the most recent push) so I was only seeing the Memref one.

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp (hunk 5 above)
Comment (joker-eph) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1093]: I believe you can write:
```
} else if (reshape->getParentBlock() =! dim->getParentBlock() && !dim.getIndex().getParentRegion()->isProperAncestor(
```

Because if the reshape and the dim are in the same block, but not the index operation, then we necessarily have the dominance structure you want.

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
@MaheshRavishankar any concern left here?

**Comment by MaheshRavishankar** [comment]
> @MaheshRavishankar any concern left here?

Nope. I dismissed my review here. Thanks for checking

**Comment by joker-eph** [comment]
Thanks! Merging then
